File: modules/tunnelguard-core/src/TunnelGuardCoreModule.ts
import { requireNativeModule } from "expo-modules-core";
import { Platform } from "react-native";

export type NativeEngineStatus = {
  state: "development-build-required" | "permission-required" | "core-not-bundled" | "idle" | "starting" | "connected" | "error";
  detail: string;
  nativeModule: boolean;
};

type TunnelGuardCoreNativeModule = {
  getEngineStatus(): Promise<NativeEngineStatus>;
  requestVpnPermission(): Promise<{ state: "granted" | "requested" }>;
  start(request: string): Promise<NativeEngineStatus>;
  stop(): NativeEngineStatus;
};

function loadNativeModule() {
  if (Platform.OS === "web") return null;
  try { return requireNativeModule<TunnelGuardCoreNativeModule>("TunnelGuardCore"); } catch { return null; }
}

const nativeModule = loadNativeModule();
const unavailable: NativeEngineStatus = { state: "development-build-required", detail: "Install an Android development build to use Heni Tech VPN native capabilities.", nativeModule: false };

export async function getEngineStatus(): Promise<NativeEngineStatus> { return nativeModule ? nativeModule.getEngineStatus() : unavailable; }
export async function requestVpnPermission() {
  if (!nativeModule) throw new Error(unavailable.detail);
  return nativeModule.requestVpnPermission();
}
export async function startEngine(request: string): Promise<NativeEngineStatus> {
  if (!nativeModule) return unavailable;
  return nativeModule.start(request);
}
export function stopEngine(): NativeEngineStatus { return nativeModule ? nativeModule.stop() : unavailable; }

export default { getEngineStatus, requestVpnPermission, startEngine, stopEngine };

File: modules/tunnelguard-core/android/src/main/java/expo/modules/tunnelguardcore/TunnelGuardCoreModule.kt
package expo.modules.tunnelguardcore

import android.net.VpnService
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class TunnelGuardCoreModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("TunnelGuardCore")

    AsyncFunction("getEngineStatus") {
      TunnelGuardVpnService.statusSnapshot()
    }

    AsyncFunction("requestVpnPermission") {
      val context = appContext.reactContext ?: throw IllegalStateException("Android context is unavailable.")
      val intent = VpnService.prepare(context)
      if (intent == null) {
        mapOf("state" to "granted")
      } else {
        val activity = appContext.currentActivity ?: throw IllegalStateException("Open TunnelGuard from an Android activity to request VPN permission.")
        activity.startActivityForResult(intent, VPN_PERMISSION_REQUEST_CODE)
        mapOf("state" to "requested")
      }
    }

    AsyncFunction("start") { config: String ->
      val context = appContext.reactContext ?: throw IllegalStateException("Android context is unavailable.")
      if (VpnService.prepare(context) != null) {
        mapOf("state" to "permission-required", "detail" to "Approve Android VPN permission before starting Heni Tech VPN.", "nativeModule" to true)
      } else {
        TunnelGuardVpnService.start(context, config)
        TunnelGuardVpnService.statusSnapshot()
      }
    }

    Function("stop") {
      val context = appContext.reactContext
      if (context != null) TunnelGuardVpnService.stop(context)
      TunnelGuardVpnService.statusSnapshot()
    }
  }

  private companion object { const val VPN_PERMISSION_REQUEST_CODE = 6241 }
}
